File: 0733-flood-fill/0733-flood-fill.cpp
// TC - O(m*n)
// SC - O(m*n)

class Solution {
    void dfs_helper(vector<vector<int>>& image, pair<int, int> index, int old_color, int new_color){
        // assign new color
        image[index.first][index.second] = new_color;

        vector<int> directions = {-1, 0, 1, 0, -1};
        for(int d=0; d<directions.size()-1; d++){
            int new_r = index.first+directions[d];
            int new_c = index.second+directions[d+1];
            if(new_r>=0 and new_c>=0 and new_r<image.size() and new_c<image[0].size() and image[new_r][new_c]==old_color)
                dfs_helper(image, make_pair(new_r, new_c), old_color, new_color);
        }
    }
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image[sr][sc]==color)
            return image;
        dfs_helper(image, {sr, sc}, image[sr][sc], color);
        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.py
# TC- O(mn)
# SC- O(mn)

class Solution:
    def dfs(self, image, rows, cols, r, c, old_color, new_color):
        image[r][c] = new_color
        dirs = [-1, 0, 1, 0, -1]
        for d in range(len(dirs)-1):
            next_r = r+dirs[d]
            next_c = c+dirs[d+1]
            if(next_r>=0 and next_r<rows and next_c>=0 and next_c<cols and image[next_r][next_c]==old_color):
                self.dfs(image, rows, cols, next_r, next_c, old_color, new_color)

    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        rows, cols = len(image), len(image[0])
        old_color = image[sr][sc]
        if old_color == color:
            return image
        self.dfs(image, rows, cols, sr, sc, old_color, color)
        return image
        